src/Service/ApplicationFactory.php

```diff
@@ -4,40 +4,51 @@
 
 namespace Laminas\Mvc\Service;
 
+use Laminas\EventManager\EventManagerInterface;
+use Laminas\EventManager\ListenerAggregateInterface;
 use Laminas\Mvc\Application;
-use Laminas\ServiceManager\Factory\FactoryInterface;
 use Psr\Container\ContainerInterface;
 
-class ApplicationFactory implements FactoryInterface
+use function assert;
+use function is_array;
+
+final class ApplicationFactory
 {
     /**
      * Create the Application service
      *
-     * Creates a Laminas\Mvc\Application service, passing it the configuration
-     * service and the service manager instance.
-     *
-     * @param  string $name
-     * @param  null|array $options
-     * @return Application
+     * Creates a Laminas\Mvc\Application service and attaches event listener aggregates as specified by
+     * configuration at `$config[Application::class]['listeners']`
      */
-    public function __invoke(ContainerInterface $container, $name, ?array $options = null)
+    public function __invoke(ContainerInterface $container): Application
     {
         $application = new Application(
             $container,
-            $container->get('EventManager'),
+            $container->get(EventManagerInterface::class),
             $container->get('Request'),
             $container->get('Response')
         );
 
-        if (! $container->has('config')) {
-            return $application;
-        }
+        self::attachApplicationListenersFromConfig($application, $container);
 
-        $em        = $application->getEventManager();
-        $listeners = $container->get('config')[Application::class]['listeners'] ?? [];
+        return $application;
+    }
+
+    public static function attachApplicationListenersFromConfig(
+        Application $application,
+        ContainerInterface $container
+    ): void {
+        $config = $container->get('config');
+        assert(is_array($config));
+
+        /** @psalm-var list<string> $listeners */
+        $listeners = $config[Application::class]['listeners'] ?? [];
```

Comment:
Presence of `listeners` configuration should be made mandatory with the default listeners moved to config.
Also to avoid problems with config merging duplicate entries should be removed.